proposals/NNNN-DXIL-Data-Scalarization.md

```diff
@@ -0,0 +1,62 @@
+# DXIL Data Scalarization
+
+* Proposal: [NNNN](NNNN-DXIL-Data-Scalarization.md)
+* Author(s): [Farzon Lotfi](https://github.com/farzonl)
+* Sponsor: [Farzon Lotfi](https://github.com/farzonl)
+* Status: **Under Consideration**
+
+## Introduction
+
+In Proposal [0009](0009-DXIL-Function-Scalarization.md) we covered scalarization
+of call instructions, and vector operations like math ops, logical ops,
+bitcasts, loads, and stores. The goal of this proposal is to present a solution
+for scalarizing data structures via data layout transformations. As a note,
+any case that results in an implicit cbuffer will not be covered by this
+proposal.
+
+## Motivation
+
+As mentioned in [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#vectors)
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+we need to be able to scalarize these types in clang DirectX backend. Without
+scalarizing the data structures and call instructions we can't generate legal
+DXIL.
+
+## Background
+
+In DXC we support five forms of data scalarization. data marked with
+`groupshared`, `static` function scope, `static` global scope, global data,
+and data in function scope. There are then subset behaviors depending on
+if we are dealing with arrays of vectors or vectors. The data
+scalarization can then be broken down into three cases:
+
+1. `static` scalar layouts
+   * vector of `M` size get scalarized into `M` elements.
+   * an array of `N` size with a vector of `M` will get scalarized into `M`
+     arrays of `N` size.
+2. `groupshared` scalar layouts
+   * vectors become arrays
+   * an array of `N` size with a vector of `M` will lower first to a 2d array
+     `N` by `M` and then a 1d.
+3. cbuffer usage for regular arrays.
+
+## Proposal
+
+In Proposal [0009](0009-DXIL-Function-Scalarization.md) we introduced the scalarizer
+pass this solves all the function scope data scalarization for us. What is left
+are the global scope cases. As mentioned in the intro the implicit cbuffer
+cases are not in scope for this work. What that leaves us with are the
+`groupshared` and `static` cases.
```

Comment:
This whole paragraph is a bit redundant with the intro. It might be sufficient to just reduce this all the way to "We need to implement a solution to handle `groupshared` and `static` globals." or something like that.